active/0006-extensibility-support-for-deployment-stacks.md

```diff
@@ -0,0 +1,857 @@
+
+Why this change is necessary becomes more clear by observing what the stacks services would be required to do with the
+current implementation. Consider this Bicep deployment:
+
+```bicep
+@secure()
+param kubeConfigParam string
+
+extension kubernetes with { // this will produce a diagnostic if the config is not provided
+  kubeConfig: kubeConfigParam
+  namespace: 'default'
+}
+
+// ... extension resources
+```
+
+In this example, the extension configuration contains a secret from a deployment parameter. A deployment parameter can
+be supplied by the user or sourced from a parent deployment. In either case, it is a user secret and thus cannot be
+persisted long term by stacks for use in future resource deletion.
+
+In the case where it's supplied by the user and without secret persistence, the only way the stacks service could
+possibly source this data again is to prompt the user for it. It is unclear how that experience would work. Does the
+user need to supply the entire old deployment's parameters back to the stack service? How do they know which version of
+inputs to supply? What if a subset of those parameters are no longer relevant? Would they only need to supply parameters
+used in configurations? How would we detect and differentiate that across an entire tree of deployments?
+
+In the case where it's supplied by a parent deployment, the problem becomes recursive. It's either provided by the user,
+is sourced from somewhere else in that parent deployment (such as another resource's outputs like a key vault or AKS
+cluster), or sourced from another parent deployment.
+
+To complicate this further, any pathway between the secret's source and consumption can have dynamic language
+expressions such as template function calls or branching. This would effectively require stacks to generate a separate
+deployment template at deletion time from the user's deployment template to process these language expressions.
+
+The stacks service cannot store user secrets, and it should not evaluate dynamic deployment template language
+expressions. The stacks service needs extension configurations to be defined in a format that is interpretable at
+deletion time that is transparent to the user and secure.
+
+### The new design from the Bicep perspective
+
+The following Bicep code illustrates the new design for extension configurations. The root deployment supplies the
+extension configuration details to the child deployment, allowing direct access to ARM resources in the root deployment
+that supply secrets.
+
+main.bicep - The root deployment
+
+```bicep
+param namespace string = 'default'
+
+// The ARM aksCluster resource can supply kubeConfigs for the AKS Kubernetes use case.
+resource aksCluster '...' = {
+  // ...
+}
+
+module extResources 'extResources.bicep' = {
+  name: 'extResourcesDeploymentName'
+  extensionConfigs: {
+    k8s: {
+      auth: {
+        kubeConfig: aksCluster.listClusterAdminCredential().kubeconfigs[0].value
```

Comment:
If I understand correctly, the essence of @anthony-c-martin suggestion (please correct me if I'm wrong) is to use a discriminator to annotate the credential source, e.g., 

```
{
  k8s: {
    // KeyVault
    kubeConfig: {
       source: 'KeyVaultSecret'
       keyVault: myKeyVault
       secret: 'mySecret' 
    }
    // List secret
    kubeConfig: {
       source: 'ResourceSecret'
       resource: myAksCluster
       credentialType: 'Admin' // or 'User'
    }
    // User provided - not supported by Stacks
    kubeConfig: {
       type: 'UserProvided'
       value: '...' // or any explicit value, can even be aksCluster.listClusterAdminCredential().kubeconfigs[0].value
    }
}
```

This avoids the need to analyze expressions, as Stacks can use the credential source type discriminator to determine whether it is safe to save the credential reference. This approach differs from the current expression-based solution.

Have we considered this? If we move forward with the expression-based solution, I don’t see much benefit in maintaining a dedicated `auth` section.

Reply:
Meeting 12/2: 
1. Change design to use this pattern for auth properties.
2. We can probably eliminate the "auth" wrapper object.
3. Check inheritance patterns against new design.
4. Future: investigate the extension providing ARM deployment engine the way to fetch credentials. For now, this will be hard coded in the deployment engine.